Need a better matching strategy

http.HandlerFunc(path, func) are very path oriented.

The ResponderConfig should generate a map of them with the key being the path. Then the wrapper HandlerFunc can search for the appropriate Responder by matching methods, headers, and body.

Probably need to rethink the `ResponderConfig.FindMatchingResponder(...)` method. Or maybe even the Yaml structure.

A lot has changed since the above comment, but this is still a problem.

For example:

```yaml
responders:
# If this appears first
- when:
    http:
      method: GET
      path: /things
    headers:
      Content-Type: text/plain
  then:
    http:
      status: 200
    headers:
      Content-Type: text/plain
    body: |
      thing1
      thing2
      thing3

# But we are trying to match this one
- when:
    http:
      method: GET
      path: /things
    headers:
      Content-Type: application/json
  then:
    http:
      status: 200
    headers:
      Content-Type: application/json
    body: |
      [
          {"Name": "thing1"},
          {"Name": "thing2"},
          {"Name": "thing3"},
      ]

```

Then we get this output:
```
go test ./test/
--- FAIL: TestUtility (0.00s)
    e2e_test.go:57: 
        expected body: '[
            {"Name": "thing1"},
            {"Name": "thing2"},
            {"Name": "thing3"},
        ]'
        actual body: 'thing1
        thing2
        thing3'
FAIL
FAIL    github.com/nicklarsennz/mock-http-response/test 0.003s
FAIL
```

--- a/mockhttp.go
+++ b/mockhttp.go
@@ -43,32 +43,30 @@ func MatchResponse(req *http.Request, config *responders.ResponderConfig) *http.
 
 	// Loop through responders which match the method and path
 	for _, responder := range config.Responders {
-		switch {
-		case responder.When.Http.Method == req.Method:
-			// logger.Logf("Match host: %s", req.Method)
-			fallthrough
-		case responder.When.Http.Path == req.URL.Path:
-			// logger.Logf("Match host: %s", req.URL.Path)
-			fallthrough
-		// Todo: case responder.When.Headers Check every responder header is in the request header?
-		case responder.When.Headers.AppearIn(req.Header):
-			// logger.Logf("Match headers")
-			fallthrough
-		case responder.When.Body == bodyString(req):
-			// logger.Logf("Match host: %s", req.URL.Path)
-			fallthrough
-		default:
-			var headers = make(http.Header)
-			for k, v := range responder.Then.Headers {
-				headers.Add(k, v)
-			}
-			return &http.Response{
-				Status:     fmt.Sprintf("%d %s", responder.Then.Http.Status, responder.Then.Http.Message),
-				StatusCode: responder.Then.Http.Status,
-				Header:     headers,
-				Body:       ioutil.NopCloser(bytes.NewBufferString(responder.Then.Body)),
-			}
 
+		// Skip if no match
+		if responder.When.Http.Method != req.Method {
+			continue
+		}
+		if responder.When.Http.Path != req.URL.Path {
+			continue
+		}
+		if !responder.When.Headers.AppearIn(req.Header) {
+			continue
+		}
+		if responder.When.Body != bodyString(req) {
+			continue
+		}
+
+		var headers = make(http.Header)
+		for k, v := range responder.Then.Headers {
+			headers.Add(k, v)
+		}
+		return &http.Response{
+			Status:     fmt.Sprintf("%d %s", responder.Then.Http.Status, responder.Then.Http.Message),
+			StatusCode: responder.Then.Http.Status,
+			Header:     headers,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(responder.Then.Body)),
 		}
 	}
 

--- a/responders/config.go
+++ b/responders/config.go
@@ -22,6 +22,7 @@ func (h Headers) AppearIn(headers http.Header) bool {
 		for _, value := range headers.Values(k) {
 			if v == value {
 				// Cool, move on to the next header
+				matches = true
 				break
 			}
 		}

--- a/test/e2e_test.go
+++ b/test/e2e_test.go
@@ -20,7 +20,7 @@ func TestUtility(t *testing.T) {
 		t.Errorf(errors.Wrap(err, "mockhttp.NewClient").Error())
 	}
 
-	req, err := http.NewRequest("GET", "http://anything/", nil)
+	req, err := http.NewRequest("GET", "http://anything/things", nil)
 	if err != nil {
 		t.Errorf(errors.Wrap(err, "http.NewRequest").Error())
 	}

--- a/test/testdata/fakes.yml
+++ b/test/testdata/fakes.yml
@@ -4,6 +4,22 @@ responders:
       method: GET
       path: /things
     headers:
+      Content-Type: text/plain
+  then:
+    http:
+      status: 200
+    headers:
+      Content-Type: text/plain
+    body: |
+      thing1
+      thing2
+      thing3
+
+- when:
+    http:
+      method: GET
+      path: /things
+    headers:
       Content-Type: application/json
   then:
     http:

